File: LeetCode/ProblemSet/Volume (300-399)/LeetCode316/LeetCode316-Solution-01.go
package leetcode316

import "sort"

func removeDuplicateLetters_01(s string) string {
	book := make([][]int, 26)
	for i := range book {
		book[i] = make([]int, 0)
	}

	for i, c := range s {
		j := int(c - 'a')
		book[j] = append(book[j], i)
	}

	length := 0
	for _, v := range book {
		if len(v) > 0 {
			length++
		}
	}

	pre, ans := -1, make([]byte, length)
	for i := 0; i < length; i++ {
		for j := 0; j < 26; j++ {
			if len(book[j]) == 0 {
				continue
			}

			ok := true
			idx := book[j][sort.Search(len(book[j]), func(idx int) bool {
				return book[j][idx] > pre
			})]

			for k := 0; k < 26; k++ {
				if len(book[k]) == 0 || j == k {
					continue
				}

				if book[k][len(book[k])-1] < idx {
					ok = false
					break
				}
			}

			if ok {
				pre = idx
				ans[i] = byte('a' + j)
				book[j] = nil
				break
			}
		}
	}

	return string(ans)
}

File: LeetCode/ProblemSet/Volume (300-399)/LeetCode316/LeetCode316-Solution-01.py
from typing import *
import bisect

class Solution:
    def removeDuplicateLetters(self, s: str) -> str:
        book = [[] for _ in range(26)]
        for i, c in enumerate(s):
            j = ord(c) - ord('a')
            book[j].append(i)

        length = sum(1 for v in book if len(v) > 0)
        pre = -1
        ans = []
        for _ in range(length):
            for i in range(0, 26):
                if not book[i]:
                    continue

                idx = book[i][bisect.bisect_right(book[i], pre)]
                ok = True
                for j in range(0, 26):
                    if not book[j] or i == j:
                        continue
                    
                    if book[j][-1] < idx:
                        ok = False
                        break

                if ok:
                    ans.append(chr(ord('a') + i))
                    pre = idx
                    book[i] = []
                    break

        return ''.join(ans)